### PR TITLE
refactor(prometheus): modularize scrape configs from prom-ksonnet

### PR DESCRIPTION
**Pull request by Duologic**
prometheus-ksonnet contains a number of very useful scrape configs, this PR aims to modularize them so they can easily
be included in other implementations.

I've taken the opportunity to order the relabel configs in the `kubernetes-pods` scrape config a bit more logically,
they don't seem to depend on each other so the order should not change the end result. I've sorted them following this
pattern:

- keep/drop targets
- handle prometheus annotations
- namespace and pod name labels

### CODE REVIEW

**Review by malcolmholmes — APPROVED** [review]
YAY! Much cleaner

File: prometheus-ksonnet/lib/kube-state-metrics.libsonnet [hunk 1]
```diff
@@ -106,4 +106,50 @@
...
+        kubernetes_sd_configs: [{
+          role: 'pod',
+          namespaces: {
+            names: [$._config.namespace],
+          },
+        }],
+
+        relabel_configs: [
+
+          // Drop anything who's service is not kube-state-metrics
```
Comment (beorn7) [inline prometheus-ksonnet/lib/kube-state-metrics.libsonnet:129]: I know it's just copied, but perhaps fix spelling, now that we are on it: who's → whose, and I find sentences easier to read if they end with a period.

File: prometheus-ksonnet/lib/kube-state-metrics.libsonnet [hunk 2]
```diff
@@ -106,4 +106,50 @@
...
+          // Rename jobs to be <namespace>/<name, from pod name label>
```
Comment (beorn7) [inline prometheus-ksonnet/lib/kube-state-metrics.libsonnet:130]: This line is a leftover and needs to go. (Again, was the case previously, but let's fix.)

File: prometheus-ksonnet/lib/node-exporter.libsonnet [hunk 3]
```diff
@@ -33,4 +33,44 @@
...
+        job_name: '%s/node-exporter' % $._config.namespace,
+        kubernetes_sd_configs: [{
+          role: 'pod',
+          namespaces: {
+            names: [$._config.namespace],
+          },
+        }],
+
+        relabel_configs: [
+          // Drop anything who's name is not node-exporter.
```
Comment (beorn7) [inline prometheus-ksonnet/lib/node-exporter.libsonnet:51]: who's → whose

File: prometheus-ksonnet/lib/node-exporter.libsonnet [hunk 4]
```diff
@@ -33,4 +33,44 @@
...
+          },
+
+          // Rename instances to be the node name.
+          {
+            source_labels: ['__meta_kubernetes_pod_node_name'],
+            action: 'replace',
+            target_label: 'instance',
+          },
+
+          // But also include the namespace as a separate label, for routing alerts
```
Comment (beorn7) [inline prometheus-ksonnet/lib/node-exporter.libsonnet:65]: End with period.

File: prometheus-ksonnet/lib/prometheus-config.libsonnet [hunk 5]
```diff
@@ -19,387 +20,44 @@ local prometheus = import 'prometheus/prometheus.libsonnet';
     },
 
     scrape_configs: [
-      {
-        job_name: 'kubernetes-pods',
-        kubernetes_sd_configs: [{
-          role: 'pod',
-        }],
+      // Grafana Labs' battle tested scrape config for scraping kubernetes pods
```
Comment (beorn7) [inline prometheus-ksonnet/lib/prometheus-config.libsonnet:23]: End with period.

File: prometheus-ksonnet/lib/prometheus-config.libsonnet [hunk 6]
```diff
@@ -19,387 +20,44 @@ local prometheus = import 'prometheus/prometheus.libsonnet';
...
-              '__meta_kubernetes_pod_container_port_name',
-            ],
-            action: 'replace',
-            separator: ':',
-            target_label: 'instance',
-          },
-        ],
-      },
+      // kube-dns does not adhere to the conventions set out by
+      // `scrape_configs.kubernetes_pods`
```
Comment (beorn7) [inline prometheus-ksonnet/lib/prometheus-config.libsonnet:27]: End with period.

File: prometheus-ksonnet/lib/prometheus-config.libsonnet [hunk 7]
```diff
@@ -19,387 +20,44 @@ local prometheus = import 'prometheus/prometheus.libsonnet';
...
+      scrape_configs.kube_dns,
 
       // This scrape config gather all kubelet metrics.
```
Comment (beorn7) [inline prometheus-ksonnet/lib/prometheus-config.libsonnet:30]: gather → gathers

File: prometheus-ksonnet/lib/prometheus-config.libsonnet [hunk 8]
```diff
@@ -19,387 +20,44 @@ local prometheus = import 'prometheus/prometheus.libsonnet';
...
-          {
-            source_labels: ['__name__'],
-            regex: 'container_(network_tcp_usage_total|network_udp_usage_total|tasks_state|cpu_load_average_10s)',
-            action: 'drop',
-          },
-        ],
-      },
+      scrape_configs.kubelet($._config.prometheus_api_server_address)
+      + (
+        // Couldn't get prometheus to validate the kublet cert for scraping, so
```
Comment (beorn7) [inline prometheus-ksonnet/lib/prometheus-config.libsonnet:33]: kublet → kubelet

File: prometheus-ksonnet/lib/prometheus-config.libsonnet [hunk 9]
```diff
@@ -19,387 +20,44 @@ local prometheus = import 'prometheus/prometheus.libsonnet';
...
+        // don't bother for now
```
Comment (beorn7) [inline prometheus-ksonnet/lib/prometheus-config.libsonnet:34]: End with period.

File: prometheus-ksonnet/lib/prometheus-config.libsonnet [hunk 10]
```diff
@@ -19,387 +20,44 @@ local prometheus = import 'prometheus/prometheus.libsonnet';
...
+        if $._config.prometheus_insecure_skip_verify
+        then scrape_configs.insecureSkipVerify
+        else {}
+      ),
+
+      // As of k8s 1.7.3, cAdvisor metrics are available via kubelet using the
+      // /metrics/cadvisor path
```
Comment (beorn7) [inline prometheus-ksonnet/lib/prometheus-config.libsonnet:41]: End with period. But perhaps we can also remove the whole comment as K8s 1.7 is ages ago.

File: prometheus/scrape_configs.libsonnet [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+  //   prometheus.io.param-<parameter> - send ?parameter=value with the scrape
+  kubernetes_pods:: {
+    job_name: 'kubernetes-pods',
+    kubernetes_sd_configs: [{
+      role: 'pod',
+    }],
+
+    relabel_configs: [
+
+      // Drop any endpoint who's pod port name does not end with metrics
```
Comment (beorn7) [inline prometheus/scrape_configs.libsonnet:25]: who's → whose.
End with period (the latter also for all the comments below).

File: prometheus/scrape_configs.libsonnet [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+        target_label: '__address__',
+        regex: '(.+?)(\\:\\d+)?;(\\d+)',
+        replacement: '$1:$3',
+      },
+
+      {
+        regex: '__meta_kubernetes_pod_annotation_prometheus_io_param_(.+)',
+        action: 'labelmap',
+        replacement: '__param_$1',
+      },
```
Comment (beorn7) [inline prometheus/scrape_configs.libsonnet:84]: Perhaps comment this stanza as well, as it is done for all the others.

**Review by beorn7 — CHANGES_REQUESTED** [review]
Just nits. I think the sorting of the relabeling config is fine.

**Review by beorn7 — APPROVED** [review]

**Comment by Duologic** [comment]
rebased.

**Pull request merged**